datajoint/fetch.py

```diff
@@ -1,71 +1,42 @@
 from collections import OrderedDict
 from collections.abc import Callable, Iterable
-from functools import wraps
+import numpy as np
 import warnings
-
 from .blob import unpack
-import numpy as np
-from datajoint import DataJointError
+from . import DataJointError
 from . import key as PRIMARY_KEY
-from . import config
-
 
-def prepare_attributes(relation, item):
-    """
-    Used by fetch.__getitem__ to deal with slices
-
-    :param relation: the relation that created the fetch object
-    :param item: the item passed to __getitem__. Can be a string, a tuple, a list, or a slice.
-
-    :return: a tuple of items to fetch, a list of the corresponding attributes
-    :raise DataJointError: if item does not match one of the datatypes above
-    """
-    if isinstance(item, str) or item is PRIMARY_KEY:
-        item = (item,)
-    elif isinstance(item, int):
-        item = (relation.heading.names[item],)
-    elif isinstance(item, slice):
-        attributes = relation.heading.names
-        start = attributes.index(item.start) if isinstance(item.start, str) else item.start
-        stop = attributes.index(item.stop) if isinstance(item.stop, str) else item.stop
-        item = attributes[slice(start, stop, item.step)]
-    try:
-        attributes = tuple(i for i in item if i is not PRIMARY_KEY)
-    except TypeError:
-        raise DataJointError("Index must be a slice, a tuple, a list, a string.")
-    return item, attributes
-
-
-def copy_first(f):
-    """
-    Decorates methods that return an altered copy of self
-    """
 
-    @wraps(f)
-    def ret(*args, **kwargs):
-        args = list(args)
-        args[0] = args[0].__class__(args[0])  # call copy constructor
-        return f(*args, **kwargs)
+class FetchBase:
 
-    return ret
+    @staticmethod
+    def _prepare_attributes(item):
+        """
+        Used by fetch.__getitem__ to deal with slices
+        :param item: the item passed to __getitem__. Can be a string, a tuple, a list, or a slice.
+        :return: a tuple of items to fetch, a list of the corresponding attributes
+        :raise DataJointError: if item does not match one of the datatypes above
+        """
+        if isinstance(item, str) or item is PRIMARY_KEY:
+            item = (item,)
+        try:
+            attributes = tuple(i for i in item if i is not PRIMARY_KEY)
+        except TypeError:
+            raise DataJointError("Index must be a sequence or a string.")
+        return item, attributes
 
 
-class Fetch(Iterable, Callable):
+class Fetch(FetchBase, Callable, Iterable):
     """
     A fetch object that handles retrieving elements from the database table.
 
     :param relation: relation the fetch object retrieves data from.
     """
 
     def __init__(self, relation):
```

Comment:
Why do we get rid of the copy constructor and copying the fetch object before it is modified?